core/src/light_protocol/query_handler.rs

```diff
@@ -0,0 +1,458 @@
+        state_manager::StateManagerTrait, SnapshotAndEpochIdRef, StateProof,
+    },
+};
+
+use super::{
+    message::{
+        GetStateEntry, GetStateRoot, Message, MsgId, RequestId,
+        StateEntry as GetStateEntryResponse, StateRoot as GetStateRootResponse,
+    },
+    Error, ErrorKind,
+};
+
+// TODO(thegaram): remove this once the `network.with_context`
+// call is moved to QueryService
+const LIGHT_PROTOCOL_ID: ProtocolId = *b"clp"; // Conflux Light Protocol
+
+const POLL_PERIOD_MS: u64 = 100;
+const MAX_POLL_TIME_MS: u64 = 1000;
+
+#[derive(Debug)]
+pub(super) enum QueryResult {
+    StateEntry(Option<Vec<u8>>),
+    StateRoot(StateRoot),
+}
+
+struct PendingRequest {
+    msg: Box<dyn Message>,
+    sender: Sender<QueryResult>,
+}
+
+pub(super) struct QueryHandler {
+    consensus: Arc<ConsensusGraph>,
+    network: Arc<NetworkService>,
+    next_request_id: RwLock<u64>,
+    peers: RwLock<HashSet<PeerId>>,
+    pending: RwLock<BTreeMap<(PeerId, RequestId), PendingRequest>>,
+}
+
+impl QueryHandler {
+    pub fn new(
+        consensus: Arc<ConsensusGraph>, network: Arc<NetworkService>,
+    ) -> Self {
+        QueryHandler {
+            consensus,
+            network,
+            next_request_id: RwLock::new(0),
+            peers: RwLock::new(HashSet::new()),
+            pending: RwLock::new(BTreeMap::new()),
+        }
+    }
+
+    // TODO(thegaram): share `send_message` and
+    // `send_message_with_throttling` with sync
+    fn send_message(
+        io: &NetworkContext, peer: PeerId, msg: &Message,
+        priority: SendQueuePriority,
+    ) -> Result<(), NetworkError>
+    {
+        QueryHandler::send_message_with_throttling(
+            io, peer, msg, priority, false,
+        )
+    }
+
+    fn send_message_with_throttling(
+        io: &NetworkContext, peer: PeerId, msg: &Message,
+        priority: SendQueuePriority, throttling_disabled: bool,
+    ) -> Result<(), NetworkError>
+    {
+        if !throttling_disabled && msg.is_size_sensitive() {
+            if let Err(e) = THROTTLING_SERVICE.read().check_throttling() {
+                debug!("Throttling failure: {:?}", e);
+                return Err(e);
+            }
+        }
+        let mut raw = Bytes::new();
+        raw.push(msg.msg_id().into());
+        raw.extend(msg.rlp_bytes().iter());
+        let _size = raw.len();
+
+        if let Err(e) = io.send(peer, raw, priority) {
+            debug!("Error sending message: {:?}", e);
+            return Err(e);
+        };
+
+        debug!(
+            "Send message({}) to {:?}",
+            msg.msg_id(),
+            io.get_peer_node_id(peer)
+        );
+        Ok(())
+    }
+
+    fn dispatch_message(
+        &self, io: &NetworkContext, peer: PeerId, msg_id: MsgId, rlp: Rlp,
+    ) -> Result<(), Error> {
+        trace!("Dispatching message: peer={:?}, msg_id={:?}", peer, msg_id);
+
+        match msg_id {
+            MsgId::GET_STATE_ROOT => self.on_get_state_root(io, peer, &rlp),
```

Comment:
Strongly recommend to handle the request/response along with the corresponding Struct definition to avoid large file. You can refer to the Messages for Full Node.